fuel-compression/src/registry/block_section.rs

```diff
@@ -0,0 +1,197 @@
+use core::fmt;
+
+use serde::{
+    ser::SerializeTuple,
+    Deserialize,
+    Serialize,
+};
+
+use super::{
+    key::Key,
+    ChangesPerTable,
+    Table,
+};
+
+/// New registrations written to a specific table.
+#[derive(Clone, PartialEq, Eq)]
+pub struct WriteTo<T: Table> {
+    /// The values are inserted starting from this key
+    pub start_key: Key<T>,
+    /// Values. inserted using incrementing ids starting from `start_key`
+    pub values: Vec<T::Type>,
+}
+
+impl<T: Table> fmt::Debug for WriteTo<T>
+where
+    T::Type: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.values.is_empty() {
+            return f.write_str("WriteTo::EMPTY");
+        }
+
+        f.debug_struct("WriteTo")
+            .field("start_key", &self.start_key)
+            .field("values", &self.values)
+            .finish()
+    }
+}
+
+impl<T: Table> WriteTo<T>
+where
+    T::Type: PartialEq,
+{
+    /// Reverse lookup.
+    /// TODO: possibly add a lookup table for this, if deemed necessary
+    pub fn lookup_value(&self, needle: &T::Type) -> Option<Key<T>> {
```

Comment:
I think we put too much business logic here with overflowing and counters. I feel like it should be part of the `fuel-core` while `fuel-compression` is responsible only for iterating over the fields and calling `next_key::<Table>(value)`. The type that implements `next_key` may be decided in which way he wants to do that.

The `count` function also sounds like something related to the exact algorithm implementation.

Reply:
I agree; if I can get the logic separated it would be really nice. I'll look into it.

Reply:
I've been trying to figure out  what parts could be moved out of fuel-vm. Most of the algorithm implementation, including this whole file, can probably be transferred.

The typed keys have to remain here, as the proc macro cannot be generic over key type. This means the implementation of `Key` has to live here.

The `count` function of the trait also has to live here. The only way to obtain the number of keys in an entity must be part of the trait. While all key allocation algorithms might not use it, many of the sensible ones do.

Reply:
Moved most of the extra stuff away.  `count` is gone as well :/